File: HighFive/HighFive/System/Double.cs
namespace System
{
    [HighFive.Convention(Member = HighFive.ConventionMember.Field | HighFive.ConventionMember.Method, Notation = HighFive.Notation.CamelCase)]
    [HighFive.External]
    [HighFive.Constructor("Number")]
    [HighFive.Reflectable]
    public struct Double : IComparable, IComparable<Double>, IEquatable<Double>, IFormattable
    {
        private extern Double(int i);

        [HighFive.Template("System.Double.max")]
        public const double MaxValue = 1.7976931348623157E+308;

        [HighFive.Template("System.Double.min")]
        public const double MinValue = -1.7976931348623157E+308;

        [HighFive.InlineConst]
        public const double Epsilon = 4.94065645841247E-324;

        [HighFive.Template("Number.NEGATIVE_INFINITY")]
        public const double NegativeInfinity = -1D / 0D;

        [HighFive.Template("Number.POSITIVE_INFINITY")]
        public const double PositiveInfinity = 1D / 0D;

        [HighFive.Template("Number.NaN")]
        public const double NaN = 0D / 0D;

        [HighFive.Template("System.Double.format({this}, {format})")]
        public extern string Format(string format);

        [HighFive.Template("System.Double.format({this}, {format}, {provider})")]
        public extern string Format(string format, IFormatProvider provider);

        public extern string ToString(int radix);

        [HighFive.Template("System.Double.format({this}, {format})")]
        public extern string ToString(string format);

        [HighFive.Template("System.Double.format({this}, {format}, {provider})")]
        public extern string ToString(string format, IFormatProvider provider);

        [HighFive.Template(Fn = "System.Double.format")]
        public override extern string ToString();

        [HighFive.Template("System.Double.format({this}, \"G\", {provider})")]
        public extern string ToString(IFormatProvider provider);

        [HighFive.Template("System.Double.parse({s})")]
        public static extern double Parse(string s);

        [HighFive.Template("HighFive.Int.parseFloat({s}, {provider})")]
        public static extern double Parse(string s, IFormatProvider provider);

        [HighFive.Template("System.Double.tryParse({s}, null, {result})")]
        public static extern bool TryParse(string s, out double result);

        [HighFive.Template("System.Double.tryParse({s}, {provider}, {result})")]
        public static extern bool TryParse(string s, IFormatProvider provider, out double result);

        public extern string ToExponential();

        public extern string ToExponential(int fractionDigits);

        public extern string ToFixed();

        public extern string ToFixed(int fractionDigits);

        public extern string ToPrecision();

        public extern string ToPrecision(int precision);

        [HighFive.Template("({d} === Number.POSITIVE_INFINITY)")]
        public static extern bool IsPositiveInfinity(double d);

        [HighFive.Template("({d} === Number.NEGATIVE_INFINITY)")]
        public static extern bool IsNegativeInfinity(double d);

        [HighFive.Template("(Math.abs({d}) === Number.POSITIVE_INFINITY)")]
        public static extern bool IsInfinity(double d);

        [HighFive.Template("isFinite({d})")]
        public static extern bool IsFinite(double d);

        [HighFive.Template("isNaN({d})")]
        public static extern bool IsNaN(double d);

        [HighFive.Template("HighFive.compare({this}, {other})")]
        public extern int CompareTo(double other);

        [HighFive.Template("HighFive.compare({this}, {obj})")]
        public extern int CompareTo(object obj);

        [HighFive.Template("{this} === {other}")]
        public extern bool Equals(double other);

        [HighFive.Template("System.Double.equals({this}, {other})")]
        public override extern bool Equals(object other);

        [HighFive.Template(Fn = "System.Double.getHashCode")]
        public override extern int GetHashCode();
    }
}

File: HighFive/HighFive/Resources/Double.js
    HighFive.define("System.Double", {
        inherits: [System.IComparable, System.IFormattable],
        statics: {
            min: -Number.MAX_VALUE,

            max: Number.MAX_VALUE,

            precision: 15,

            $number: true,

            $is: function (instance) {
                return typeof (instance) === "number";
            },

            getDefaultValue: function () {
                return 0;
            },

            parse: function (s, provider) {
                return HighFive.Int.parseFloat(s, provider);
            },

            tryParse: function (s, provider, result) {
                return HighFive.Int.tryParseFloat(s, provider, result);
            },

            format: function (number, format, provider) {
                return HighFive.Int.format(number, format || 'G', provider, System.Double);
            },

            equals: function (v1, v2) {
                if (HighFive.is(v1, System.Double) && HighFive.is(v2, System.Double)) {
                    v1 = HighFive.unbox(v1, true);
                    v2 = HighFive.unbox(v2, true);

                    if (isNaN(v1) && isNaN(v2)) {
                        return true;
                    }

                    return v1 === v2;
                }

                return false;
            },

            equalsT: function (v1, v2) {
                return HighFive.unbox(v1, true) === HighFive.unbox(v2, true);
            },

            getHashCode: function (v) {
                var value = HighFive.unbox(v, true);

                if (value === 0) {
                    return 0;
                }

                if (value === Number.POSITIVE_INFINITY) {
                    return 0x7FF00000;
                }

                if (value === Number.NEGATIVE_INFINITY) {
                    return 0xFFF00000;
                }

                return HighFive.getHashCode(value.toExponential());
            }
        }
    });

    System.Double.$kind = "";
    HighFive.Class.addExtend(System.Double, [System.IComparable$1(System.Double), System.IEquatable$1(System.Double)]);

    HighFive.define("System.Single", {
        inherits: [System.IComparable, System.IFormattable],
        statics: {
            min: -3.40282346638528859e+38,

            max: 3.40282346638528859e+38,

            precision: 7,

            $number: true,

            $is: System.Double.$is,

            getDefaultValue: System.Double.getDefaultValue,

            parse: System.Double.parse,

            tryParse: System.Double.tryParse,

            format: function (number, format, provider) {
                return HighFive.Int.format(number, format || 'G', provider, System.Single);
            },

            equals: function (v1, v2) {
                if (HighFive.is(v1, System.Single) && HighFive.is(v2, System.Single)) {
                    v1 = HighFive.unbox(v1, true);
                    v2 = HighFive.unbox(v2, true);

                    if (isNaN(v1) && isNaN(v2)) {
                        return true;
                    }

                    return v1 === v2;
                }

                return false;
            },

            equalsT: function (v1, v2) {
                return HighFive.unbox(v1, true) === HighFive.unbox(v2, true);
            },

            getHashCode: System.Double.getHashCode
        }
    });

    System.Single.$kind = "";
    HighFive.Class.addExtend(System.Single, [System.IComparable$1(System.Single), System.IEquatable$1(System.Single)]);
